### ISSUE TITLE
Enhancement: automatic reduction of advance step in ReactorNet simulations

### ISSUE THREAD

**Issue by ischoegl** [Cantera/cantera#628]
### Cantera version
2.5.0a2
### Current Behavior
There is no option for an automatic reduction of time increments between outputs for `ReactorNet.advance` when states rapidly change (e.g. example simulations for `IdealGasConstPressureReactor` in `reactor1.py`)
### Enhancement
An automatic reduction of time steps ensures that processes are properly resolved without having to guess an appropriate step size.
### Steps to reproduce
1. run example script `reactor1.py`, where the maximum temperature difference exceeds 500 K, i.e.
```
    t [s]      T [K]     P [Pa]       u [J/kg]
 ...
 3.100e-04   1207.525 101325.000   5.456362e+05
 3.200e-04   1718.185 101325.000   3.697981e+05
...
```
2. At `t = 3.100e-4`, a maximum temperature difference of ~50K between outputs requires a time advance of 1.25e-06 (8-fold reduction) at the given conditions.
3. For simulations of ignition processes, delays change over several orders of magnitude. Using small time steps by default may create unnecessarily large output files; the determination of an appropriate time step may involve trial and error.

**Comment by bryanwweber** [comment]
Hi @ischoegl! How is this different from the `step()` method that already exists? The CVODE solver is already an adaptive time-step integrator, and the `step()` method takes one internal time step, after which you can query the state of the reactor.

**Comment by ischoegl** [comment]
@bryanwweber  ... `step()` just takes one internal step within CVODE. `advance(time)` will advance to `time` no matter what (using many internal CVODE steps). The pull request adds new functionality that currently does not exist (but uses existing CVODE infrastructure). The goal was to (automatically) adjust the time increment between ouputs (i.e. states written to a file).
PS: Querying the reactor after `step()` (especially from python) will negate speed advantages of C++ over python. The actual step size is further controlled internally by CVODE's adaptive time-step integrator, and thus not easily manipulated.

**Comment by skyreflectedinmirrors** [comment]
Bryan, I think the suggested change is that the user would be able to set
an absolute or relative tolerance on the change in specific solution
components (e.g., 50K temperature intervals) that if exceeded would trigger
an intermediate output state.

I'm generally for this change as it would provide more granular (and
frankly better) output control than either the pure adaptive time step of
`step()` or the constant time step of `advance()`. I've have to implement
similar methods for plotting of the solution (using `step()`) to thin the
sample rate and make the markers easier to see.

My only question is where this should be implemented: I could see it being
triggered by optional arguments `atol` and `rtol` given to the cython
advance call, but then we would either have to return multiple solution
points (for the intermediate output points reached within an `advance`) or
return the state at a time possibly less than the one the user asked for
(breaking backwards compatibly and being fairly unintuitive).

I would suggest a new cython method (perhaps `advance_with()`?) or the
optional keyword arguments for `advance` as a solution. This would return a
`SolutionArray` with the given outputs, and times stored within as an extra
`time` variable.

**Comment by ischoegl** [comment]
@arghdos ... that's exactly what I tried to implement. I agree that there are likely better (i.e. less awkward) implementations, but I tried to refrain from changing core code when proposing this behaviour. The pull request #629 can be easily changed to accommodate an improved interface.

**Comment by ischoegl** [comment]
To be more specific: it is fairly straight-forward to pass optional arguments to `advance`, and have default tolerances that do not break backwards compatibility. As @arghdos stated, the key issue is the function potentially not returning the desired end state; my approach would be to switch from `void` to returning the actual end time as a `double`. What I implemented in #629 is essentially `advance_with` albeit named differently and not returning a `SolutionArray` (I'd be happy to change function names or return values).

**Comment by ischoegl** [comment]
I just renamed some functions in #629 to clarify (using `advance_towards(time)`, `atol_advance`, etc.)
